### PR TITLE
Check docs and run clippy on PRs

### PR DESCRIPTION
**Pull request by jsdw**
Adapt the CI from Soketto/Desub to run here. Main changes:
- Jobs run in parallel (and it's more visible at a glance what failed).
- Caching. Seems to speed up some of the checks quite nicely at least.
- Add clippy check (and fix clippy lints).
- Add doc link test (and fix doc things).

### CODE REVIEW

**Review by ascjones — APPROVED** [review]
Nice and fast CI checks :rocket:

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -2,39 +2,137 @@ name: Rust
...
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
 
-    - name: test
-      run: cargo test --workspace --verbose
+      - name: Run clippy
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: -- -D warnings
```
Comment (niklasad1) [inline .github/workflows/rust.yml:138]: all targets?!

```suggestion
          args: --all-targets -- -D warnings
```

**Review by niklasad1 — APPROVED** [review]

File: .github/workflows/rust.yml (hunk 1 above)
Comment (jsdw) [reply .github/workflows/rust.yml:138]: Ooh good catch! That picked up some more lint issues to fix too :D :'(

Comment (niklasad1) [reply .github/workflows/rust.yml:138]: maybe postpone it to another PR I suspect that it will plenty of work :)

Comment (jsdw) [reply .github/workflows/rust.yml]: Oh, I've just finished fixing them so it's all gravy :)

**Pull request merged**